File: backend/TradeX.Core/Models/Position.cs
using TradeX.Core.Abstractions;
using TradeX.Core.Enums;
using TradeX.Core.Events;
using TradeX.Core.Interfaces;

namespace TradeX.Core.Models;

/// <summary>
/// 持仓聚合根。生产代码使用 <see cref="UpdateMarketPrice"/> / <see cref="Close"/> 而非直接 setter。
///
/// 状态机：
/// <code>
/// Open ──UpdateMarketPrice──► Open (持续刷新 CurrentPrice/UnrealizedPnl)
///      ──Close──► Closed (终态)
/// </code>
/// </summary>
public class Position : AggregateRoot, IVersioned
{
    // EF Core 无参构造函数（公开以兼容现有代码）
    public Position() { }

    /// <summary>工厂方法：开仓。</summary>
    public static Position Open(
        Guid traderId, Guid exchangeId, Guid strategyId,
        string pair, decimal quantity, decimal entryPrice)
    {
        var pos = new Position
        {
            TraderId = traderId,
            ExchangeId = exchangeId,
            StrategyId = strategyId,
            Pair = pair,
            Quantity = quantity,
            EntryPrice = entryPrice,
            CurrentPrice = entryPrice
        };
        pos.AddDomainEvent(new PositionOpenedDomainEvent(
            pos.Id, traderId, strategyId, pair, quantity, entryPrice));
        return pos;
    }

    public Guid Id { get; init; } = Guid.NewGuid();
    public Guid TraderId { get; init; }
    public Guid ExchangeId { get; init; }
    public Guid StrategyId { get; init; }
    /// <summary>
    /// 开仓买单的订单 Id。作为"成交→持仓"投影的幂等键：同一笔成交无论被实盘同步路径
    /// 还是对账器恢复路径重复处理，凭此唯一约束保证只开一条持仓。
    /// </summary>
    public Guid? OpeningOrderId { get; set; }
    public string Pair { get; set; } = string.Empty;
    public decimal Quantity { get; set; }
    public decimal EntryPrice { get; set; }
    public decimal CurrentPrice { get; set; }
    public decimal UnrealizedPnl { get; set; }
    public decimal RealizedPnl { get; set; }
    public PositionStatus Status { get; set; } = PositionStatus.Open;
    public DateTime OpenedAtUtc { get; init; } = DateTime.UtcNow;
    public DateTime? ClosedAtUtc { get; set; }
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    /// <summary>乐观并发控制版本号，由 VersionInterceptor 在保存前自动刷新。</summary>
    public Guid Version { get; set; } = Guid.NewGuid();

    // ─────────────── 领域方法 ───────────────

    /// <summary>
    /// 用当前市价刷新持仓（仅对 Open 状态生效）。
    /// 计算未实现盈亏 = (CurrentPrice - EntryPrice) * Quantity。
    /// </summary>
    public void UpdateMarketPrice(decimal currentPrice)
    {
        if (currentPrice < 0)
            throw new ArgumentOutOfRangeException(nameof(currentPrice), "价格不能为负");
        if (Status != PositionStatus.Open)
            throw new InvalidOperationException($"持仓 {Id} 已 {Status}，不能更新价格");

        CurrentPrice = currentPrice;
        UnrealizedPnl = (currentPrice - EntryPrice) * Quantity;
        UpdatedAt = DateTime.UtcNow;
    }

    /// <summary>
    /// 关闭持仓，记录实现盈亏并清空未实现盈亏。Open → Closed 转换。
    /// </summary>
    public void Close(decimal exitPrice)
    {
        if (Status != PositionStatus.Open)
            throw new InvalidOperationException($"持仓 {Id} 已 {Status}，不能再次关闭");

        RealizedPnl = (exitPrice - EntryPrice) * Quantity;
        UnrealizedPnl = 0;
        CurrentPrice = exitPrice;
        Status = PositionStatus.Closed;
        var now = DateTime.UtcNow;
        ClosedAtUtc = now;
        UpdatedAt = now;

        AddDomainEvent(new PositionClosedDomainEvent(
            Id, TraderId, Pair, Quantity, EntryPrice, exitPrice, RealizedPnl));
    }
}

File: backend-go/internal/domain/position.go
package domain

import (
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

// Position 持仓聚合根。字段与状态机对齐 C# TradeX.Core.Models.Position 及 positions 表列。
// 状态机：Open ──UpdateMarketPrice──► Open；──Close──► Closed（终态）。
type Position struct {
	ID             uuid.UUID       `json:"id"`
	TraderID       uuid.UUID       `json:"traderId"`
	ExchangeID     uuid.UUID       `json:"exchangeId"`
	StrategyID     uuid.UUID       `json:"strategyId"`
	OpeningOrderID *uuid.UUID      `json:"openingOrderId,omitempty"`
	Pair           string          `json:"pair"`
	Quantity       decimal.Decimal `json:"quantity"`
	EntryPrice     decimal.Decimal `json:"entryPrice"`
	CurrentPrice   decimal.Decimal `json:"currentPrice"`
	UnrealizedPnl  decimal.Decimal `json:"unrealizedPnl"`
	RealizedPnl    decimal.Decimal `json:"realizedPnl"`
	Status         PositionStatus  `json:"status"`
	OpenedAtUtc    time.Time       `json:"openedAtUtc"`
	ClosedAtUtc    *time.Time      `json:"closedAtUtc,omitempty"`
	UpdatedAt      time.Time       `json:"updatedAt"`
	Version        uuid.UUID       `json:"version"`
}

// OpenPosition 工厂方法：开仓（对应 C# Position.Open）。
func OpenPosition(traderID, exchangeID, strategyID uuid.UUID, pair string, quantity, entryPrice decimal.Decimal) *Position {
	now := time.Now().UTC()
	return &Position{
		ID:           uuid.New(),
		TraderID:     traderID,
		ExchangeID:   exchangeID,
		StrategyID:   strategyID,
		Pair:         pair,
		Quantity:     quantity,
		EntryPrice:   entryPrice,
		CurrentPrice: entryPrice,
		Status:       PositionStatusOpen,
		OpenedAtUtc:  now,
		UpdatedAt:    now,
		Version:      uuid.New(),
	}
}

// UpdateMarketPrice 用当前市价刷新持仓（仅 Open 生效）。UnrealizedPnl=(price-entry)*qty。
func (p *Position) UpdateMarketPrice(currentPrice decimal.Decimal) error {
	if currentPrice.IsNegative() {
		return fmt.Errorf("价格不能为负")
	}
	if p.Status != PositionStatusOpen {
		return fmt.Errorf("持仓 %s 已 %s，不能更新价格", p.ID, p.Status)
	}
	p.CurrentPrice = currentPrice
	p.UnrealizedPnl = currentPrice.Sub(p.EntryPrice).Mul(p.Quantity)
	p.UpdatedAt = time.Now().UTC()
	return nil
}

// Close 关闭持仓，记录实现盈亏（Open → Closed）。
func (p *Position) Close(exitPrice decimal.Decimal) error {
	if p.Status != PositionStatusOpen {
		return fmt.Errorf("持仓 %s 已 %s，不能再次关闭", p.ID, p.Status)
	}
	p.RealizedPnl = exitPrice.Sub(p.EntryPrice).Mul(p.Quantity)
	p.UnrealizedPnl = decimal.Zero
	p.CurrentPrice = exitPrice
	p.Status = PositionStatusClosed
	now := time.Now().UTC()
	p.ClosedAtUtc = &now
	p.UpdatedAt = now
	return nil
}
